### PR TITLE
BUG:  chokes on pd.DatetimeTZDtype if there are no rows.

### LINKED ISSUE #59126: BUG: pivot_table chokes on pd.DatetimeTZDtype if there are no rows.
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
df = pd.DataFrame({'timestamp': [pd.Timestamp('2024-01-01T00:00', tz='UTC')], 'category': ['A'], 'value': [100]})
df[df.category == "C"].pivot_table(index="category", columns="value", values="timestamp")
```

### Issue Description

This raises an error `TypeError: Cannot interpret 'datetime64[ns, UTC]' as a data type`

### Expected Behavior

It should return an empty DataFrame.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : d9cdd2ee5a58015ef6f4d15c7226110c9aab8140
python                : 3.12.3.final.0
python-bits           : 64
OS                    : Darwin
OS-release            : 23.5.0
Version               : Darwin Kernel Version 23.5.0: Wed May  1 20:14:38 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6020
machine               : arm64
processor             : arm
byteorder             : little
LC_ALL                : None
LANG                  : fr_FR.UTF-8
LOCALE                : fr_FR.UTF-8

pandas                : 2.2.2
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 69.5.1
pip                   : 24.0
Cython                : None
pytest                : 8.2.1
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : 3.2.0
lxml.etree            : 5.2.1
html5lib              : None
pymysql               : None
psycopg2              : 2.9.9
jinja2                : 3.1.4
IPython               : 8.23.0
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.3
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : 2024.5.0
gcsfs                 : None
matplotlib            : 3.8.4
numba                 : None
numexpr               : None
odfpy                 : None
openpyxl              : 3.1.0
pandas_gbq            : None
pyarrow               : 16.1.0
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : 2024.5.0
scipy                 : None
sqlalchemy            : 2.0.30
tables                : None
tabulate              : 0.9.0
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None

</details>

### PR DESCRIPTION
**Pull request by pcorpet**
This is a follow up to #41875

- [X] closes #59126
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by mroeschke** [comment]
Is there an issue open for this?

**Comment by pcorpet** [comment]
I'd welcome some help to understand where/how to add a proper test. I have added a simple repro (3 lines) in the bug

**Comment by mroeschke** [comment]
If you could add a unit test in `pandas/tests/reshape/test_pivot.py` that mimics your reproducer that would be great

File: pandas/core/reshape/reshape.py [hunk 1]
```diff
@@ -291,7 +291,12 @@ def get_new_values(self, values, fill_value=None):
         # if our mask is all True, then we can use our existing dtype
         if mask_all:
             dtype = values.dtype
-            new_values = np.empty(result_shape, dtype=dtype)
+            if isinstance(dtype, ExtensionDtype):
```
Comment (mroeschke) [inline pandas/core/reshape/reshape.py:294]: Could you combine this branch with the one below?  They are very similar

File: doc/source/whatsnew/v3.0.0.rst [hunk 2]
```diff
@@ -594,6 +594,7 @@ Reshaping
 ^^^^^^^^^
 - Bug in :meth:`DataFrame.join` inconsistently setting result index name (:issue:`55815`)
 - Bug in :meth:`DataFrame.unstack` producing incorrect results when ``sort=False`` (:issue:`54987`, :issue:`55516`)
+- Bug in :meth:`DataFrame.unstack` producing incorrect results when manipulating empty DataFrame using dtype extensions (:issue:`59123`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v3.0.0.rst:597]:
```suggestion
- Bug in :meth:`DataFrame.unstack` producing incorrect results when manipulating empty :class:`DataFrame` with an :class:`ExtentionDtype` (:issue:`59123`)
```

File: pandas/tests/reshape/test_pivot.py [hunk 3]
```diff
@@ -2769,3 +2769,18 @@ def test_unstack_copy(self, m):
         result = df.unstack(sort=False)
         result.iloc[0, 0] = -1
         tm.assert_frame_equal(df, df_orig)
+
+    @pytest.mark.parametrize("tz", [None, "UTC"])
+    def test_pivot_empty_with_datetime(self, tz):
+        # GH#59126
+        df = DataFrame(
+            {
+                "timestamp": [pd.Timestamp("2024-01-01T00:00", tz=tz)],
```
Comment (mroeschke) [inline pandas/tests/reshape/test_pivot.py:2778]: If your create the columns with `pd.Series(None, dtype=...)` you should get an empty `df` with the dtypes already (without having to do that filter in the next line)

File: doc/source/whatsnew/v3.0.0.rst (hunk 2 above)
Comment (pcorpet) [reply doc/source/whatsnew/v3.0.0.rst]: Thanks. Done.

File: pandas/core/reshape/reshape.py (hunk 1 above)
Comment (pcorpet) [reply pandas/core/reshape/reshape.py]: I'm not sure it's way better, but I've tried.

File: pandas/tests/reshape/test_pivot.py (hunk 3 above)
Comment (pcorpet) [reply pandas/tests/reshape/test_pivot.py]: Yes, thanks!

Comment (pcorpet) [reply pandas/tests/reshape/test_pivot.py:2778]: Actually I tried and keep getting the error before even testing:
https://github.com/pandas-dev/pandas/actions/runs/9713703751/job/26811065289

I guess we should also try to fix the construction of empty Series, but this is not the bug I'm after here.

Comment (mroeschke) [reply pandas/tests/reshape/test_pivot.py:2778]: > "timestamp": pd.Series([], pd.DatetimeTZDtype(tz="UTC")),

The second argument positionally is `index`, not `dtype`

File: pandas/core/reshape/reshape.py [hunk 4]
```diff
@@ -291,18 +291,19 @@ def get_new_values(self, values, fill_value=None):
         # if our mask is all True, then we can use our existing dtype
         if mask_all:
             dtype = values.dtype
-            new_values = np.empty(result_shape, dtype=dtype)
```
Comment (mroeschke) [inline pandas/core/reshape/reshape.py:294]: I think you can remove the 3 lines above this one too

File: pandas/tests/reshape/test_pivot.py (hunk 3 above)
Comment (pcorpet) [reply pandas/tests/reshape/test_pivot.py]: Riiight, sorry I was sloppy.

File: pandas/core/reshape/reshape.py (hunk 4 above)
Comment (pcorpet) [reply pandas/core/reshape/reshape.py:294]: Indeed.

File: pandas/tests/reshape/test_pivot.py [hunk 5]
```diff
@@ -2769,3 +2769,17 @@ def test_unstack_copy(self, m):
...
         result.iloc[0, 0] = -1
         tm.assert_frame_equal(df, df_orig)
+
+    def test_pivot_empty_with_datetime(self):
+        # GH#59126
+        df = DataFrame(
+            {
+                "timestamp": pd.Series([], dtype=pd.DatetimeTZDtype(tz="UTC")),
+                "category": pd.Series([], dtype=str),
+                "value": pd.Series([], dtype=str),
```
Comment (mroeschke) [inline pandas/tests/reshape/test_pivot.py:2779]:
```suggestion
                "timestamp": Series([], dtype=pd.DatetimeTZDtype(tz="UTC")),
                "category": Series([], dtype=str),
                "value": Series([], dtype=str),
```

Comment (pcorpet) [reply pandas/tests/reshape/test_pivot.py]: Done.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @pcorpet

**Pull request merged**